kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/Protocol.java

```diff
@@ -105,23 +141,261 @@ public Row toRow() {
     Map<Integer, Object> protocolMap = new HashMap<>();
     protocolMap.put(0, minReaderVersion);
     protocolMap.put(1, minWriterVersion);
-    protocolMap.put(2, stringArrayValue(readerFeatures));
-    protocolMap.put(3, stringArrayValue(writerFeatures));
+    protocolMap.put(2, stringArrayValue(new ArrayList<>(readerFeatures)));
+    protocolMap.put(3, stringArrayValue(new ArrayList<>(writerFeatures)));
 
     return new GenericRow(Protocol.FULL_SCHEMA, protocolMap);
   }
 
+  /////////////////////////////////////////////////////////////////////////////////////////////////
+  /// Public methods related to table features interaction with the protocol                    ///
+  /////////////////////////////////////////////////////////////////////////////////////////////////
+  /**
+   * Get the set of features that are implicitly enabled by the protocol. Features are implicitly
+   * enabled if they reader and/or writer version is less than the versions that support the
+   * explicit features specifying in `readerFeatures` and `writerFeatures` sets.
+   */
+  public Set<TableFeature> getImplicitlyEnabledFeatures() {
+    if (supportsReaderFeatures && supportsWriterFeatures) {
+      return emptySet();
+    } else {
+      return TableFeatures.TABLE_FEATURES.stream()
+          .filter(f -> !supportsReaderFeatures && f.minReaderVersion() <= minReaderVersion)
+          .filter(f -> !supportsWriterFeatures && f.minWriterVersion() <= minWriterVersion)
+          .collect(Collectors.toSet());
+    }
+  }
+
+  /**
+   * Get the set of features that are explicitly enabled by the protocol. Features are explicitly
+   * enabled if they are present in the `readerFeatures` and/or `writerFeatures` sets.
+   */
+  public Set<TableFeature> getExplicitlyEnabledFeatures() {
+    // TODO: Should we throw an exception if we encounter a feature that is not known to Kernel yet?
```

Comment:
That wouldn't happen in this method, right? That should probably happen when we construct the protocol?

i.e. if Delta Kernel is pointed to a table and the table has Delta Table readerWriter Feature `foo` that Kernel doesn't support, we should throw?

and again on the write path we should check if there's a writer only feature that we don't support?

Let me know if I'm missing something -- I'd figure we were already doing that today